keps/140-scenario-based-simulation/README.md

```diff
@@ -43,368 +64,539 @@ Therefore, this kep proposes to define the scenario **as CRD**. All clients, inc
 
 ```go
-type Scenario struct {
-  metav1.TypeMeta 
-  metav1.ObjectMeta // non namespaced.
-
-  Spec ScenarioSpec
-  Status ScenarioStatus
+// ScenarioSpec defines the desired state of Scenario
+type ScenarioSpec struct {
+	// Operations field has all operations for a scenario.
+	// Also, you can add a new operation while the scenario runs.
+	//
+	// +patchMergeKey=ID
+	// +patchStrategy=merge
+	Operations []*ScenarioOperation `json:"operations"`
+
+	// Controllers have the configuration for controllers working with simulation.
+	Controllers *Controllers `json:"controllers"`
 }
 
-type ScenarioSpec struct {
-  // Events field has all operations for a scenario.
-  // Also you can add new events during the scenario is running.
-  //
-  // +patchMergeKey=ID
-  // +patchStrategy=merge
-  Events   []ScenarioEvent
+type Controllers struct {
+	// PreparingControllers is a list of controllers that should be run before SimulatedControllers.
+	// They will run in parallel.
+	// 
+	// It's an optional field. 
+	// All controllers registered in the simulator will be enabled automatically. (except controllers set in Simulate.)
+	// So, you need to configure it only when you want to disable some controllers enabled by default.
+	//
+	// +optional
+	PreparingControllers	*ControllerSet `json:"preparingControllers`
+	// SimulatedControllers is a list of controllers that are the target of this simulation.
+	// These are run one by one in the same order specified in Enabled field.
+	// 
+	// It's a required field; no controllers will be enabled automatically.
+	SimulatedControllers  *ControllerSet `json:"simulatedControllers"`
 }
 
-type ScenarioEvent struct {
-  // ID for this event. Normally, the system sets this field for you.
-  ID string
-  // Step indicates the step at which the event occurs.
-  Step ScenarioStep
-  // Operation describes which operation this event wants to do.
-  // Only "Create", "Patch", "Delete", "Done" are valid operations in ScenarioEvent.
-  Operation OperationType
-
-  // One of the following four fields must be specified.
-  // If more than one is specified or if all are empty, the event is invalid and the scenario will fail.
-
-  // CreateOperation is the operation to create new resource.
-  // When use CreateOperation, Operation should be "Create".
-  CreateOperation *CreateOperation
-  // PatchOperation is the operation to patch a resource.
-  // When use PatchOperation, Operation should be "Patch".
-  PatchOperation *PatchOperation
-  // DeleteOperation indicates the operation to delete a resource.
-  // When use DeleteOperation, Operation should be "Delete".
-  DeleteOperation *DeleteOperation
-  // DoneOperation indicates the operation to mark the scenario as DONE.
-  // When use DoneOperation, Operation should be "Done".
-  DoneOperation   *DoneOperation
+type ControllerSet struct {
+	// Enabled specifies controllers that should be enabled.
+	// +listType=atomic
+	Enabled  []Controller `json:"enabled"`
```

Comment:
The order of Controllers affects the execution orders of SimulatedControllers?? If so, we should explain this explicitly.

Reply:
Yea, it's described in L95.

Reply:
oh, I missed it. Thanks.